### DISCUSSION TITLE
Best way to add a computed value to a auth request

### DISCUSSION THREAD

**Discussion by ghostdevv** [pocketbase/pocketbase#5828] [Q&A] [answered]
I want to check if a user has a record in the submissions collection, and attach it as a boolean `hasSubmission` to the user record returned after authenticating. Currently in my TS app I'm just making a follow up request something like this:

```js
await locals.pb
	.collection('submissions')
	.getFirstListItem('', { fields: 'id' })
	.then(() => true)
	.catch(() => false)
```

This is kinda inefficient as it's a serverless app, so I'd like to avoid adding another request. My current ideas are:

- Maybe there is something I missed that allows doing this already
- Add the `hasSubmission` bool to the record in a `OnRecordViewRequest` hook

Ideally it's something that follows the api rules (for the submissions collection), since the `view rule` is rather long and I don't want to duplicate that logic. Any suggestions?

**Comment by ganigeorgiev** [comment, marked as answer]
There is no easy way to do that without using the hooks. Note that custom record fields will be serialized only when `record.WithUnknownData(true)` is explicitly set (_the method is renamed to `WithCustomData` with v0.23.0-rc_).

Depending on your use case an easier alternative could be to use [back-relation](https://pocketbase.io/docs/working-with-relations/#back-relations) `expand`. For example, assuming that the user `relation` field in your `submissions` collection is named `user` this could look like:

```js
await pb.collection("users").authWithPassword("test@example.com", "1234567890", {
    expand: "submissions_via_user",
})
```

**Reply by ghostdevv** [reply]
Ah thanks for confirming that! I'm not able to use a back relation in this case unfortunately, as there are multiple fields in the `submissions` collection that give a user view access

**Reply by dschissler** [reply]
You can create fancy OR collection API logic for any of that.  For development try making it wide open first to see if you can get it to work.  Then add each case one by one with `||` until you've allowed all of them.

**Reply by ghostdevv** [reply]
What do you mean? @dschissler

**Reply by dschissler** [reply]
@ghostdevv Here is my general strategy for developing expands and API Rules while still in the early prototyping phase:

1) Unlock the API rules for the collection for everyone, for all collections.  2) Add the expand that you want in the API call. 3) Check that the expanded data arrives. 4) Lock the API Rule down again by adding a pass conditions to the primary collection (of the API call) or to the expanded collection.  5) Check that the API call returns data and with the expanded data as well for some condition. 6) Continue adding extra pass conditions to the expanded collection API rule until it works for all separate possible cases (each joined via the `||` operator, wrapped in parenthesis).  7) Optimize the logic as people do.

While developing you can also make direct calls to the expanded collection via API to see if it will return data with the API Rules that you allowed for it.  These rules will be used while using that collection as an expand.

To clarify, the expand is like if you were to make a API call to that collection itself.  So it processes all of that over again according to the rules.  If you were to require this part to be ultra efficient then you should probably make a custom endpoint for that and checking permission in some other way.  Otherwise if it's not a critical point of the application then it shouldn't matter at all (and might not even anyways).

Said another way, each collection API rule stands on its own whether called directly or while being expanded.

So perhaps each collection should be checking something against some group membership table with `@request.auth.id` against some table joins (probably) using the `?=` operator.  Search the previous discussions for this.

https://pocketbase.io/docs/api-rules-and-filters/
